records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,135 @@
+the database, it commits each of these individually.
+3. The notification is then sent to the appropriate queue, depending on its
+type, scheduling and used key.
+4. In the case of an email that was requested through the API, the email
+will be sent to the AWS SES provider (the only supported at the time of
+this writing).
+5. As multiples emails are sent and some receive a response receipt, multiple
+updates on the database are performed.
+6. In the meantime, numerous count operations are performed to track the
+current delivered emails and various metrics.
+7. As there are more and more commits performed on the database, for each and
+every notification, the database triggers an autovacuum action to clean up the
+previous inserted rows by design of its
+[Multiversion Concurrency Control](https://www.postgresql.org/docs/11/mvcc-intro.html).
+8. The database performance decreases as autovacuum is running and slows down
+the concurrent insertion, update and reading of the `notifications` table
+records.
+9. As the *notification-admin* and *notification-api* components are allegedly
+slowing down due to a slower database, the healthcheck response[1] time gets higher
+than normal and the ALB stop serving these instances, effectively making these
+to stop receiving any request. It affects all instances of the API and Admin
+which both heavily relies on calls to the same database.
+10. As the database slowed down to answer the API, the latter does not stop
+receiving and processing notifications. It continues to makes calls to the database
+and consumes more database connections as previous ones are still busy.
+11. Errors of maximum number of database connections getting reached start
+to appear at that point.
+12. The API instances are down as these are detached from the load balancers,
+the database is overloaded with connections and can't take any more. **We are 
+losing notifications** because we don't save these when the API is unavailable.
+
+[1] The ALB makes health checks on Notify's stack components to make sure these
+are eligible for traffic, but they call a `_status?simple=true` endpoint that makes
+sure a database call is not included as part of the check. This means the situation
+might be more dire as the instance overall slowed down so much due to the database
+might lock the process due to its unavailability.
+
+_TODO: Add sequence or flow chart_
+
+## Options
+
+### Stream the notifications save
+
+Upon receiving a notification from the API, do not save it in the current
+*notification-api* process but defer it to a celery task.
+
+### Insert the notifications as a batch
+
+As we streamline the notification save with the previous option and got
+these under Celery land, we can now batch these easily. There is a 
+[celery-batch](https://pypi.org/project/celery-batches/) project which 
+exists that might offer us the option to easily achieve this with minimal
+setup.
+
+### Tweaking autovacuum options
```

Comment:
I feel anything that we would try to control or manage a built in database mechanism of cleaning is a risky move, and a temporary solution at best. I'd prefer we look at reviewing the architecture of the solution.

Reply:
This is our best mitigation for short term, as we are losing notifications. A combination of manual run (but keeping automatic runs) + options tweaking could be beneficial. I am working on recommendations from the AWS expert on this. We also now have a quasi similar production-like environment in staging that we can use to test the configurations first on knowing the risk of these while [we run load/stress tests in these](https://github.com/cds-snc/notification-api/tree/master/tests-perf/locust).

Although, indeed, I agree that solving the design is the long term solution, to me at least, but I wouldn't exclude a more proper tweaking of these to accommodate a burst profile like Notify has on long-term with careful grooming in staging.

Reply:
Don't underestimate just provisioning larger instances to buy time. I am not a big fan of buying a way out of an efficiency problem, but I believe the cost difference from existing to r6large or r6xlarge is minimal when compared to costly developer time. It could be considered kicking the can down the road, but with a solid plan to mitigate and reduce unnecessary load, it can be a worthwhile investment.

Reply:
@dorianjp we already have a r6large in prod and staging actually. This is not a problem we're going to easily be able to buy our way out of unfortunately